### PR TITLE
Custom compression codecs

### PR DESCRIPTION
**Pull request by alesapin**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Category (leave one):
- New Feature

Short description (up to few sentences):
Add custom compression codecs.

### CODE REVIEW

**Comment by alesapin** [comment]
- [ ] Read own code
- [ ] Add tests (likely integration) with non default compression codecs for tables and for network connection

**Comment by alesapin** [comment]
Failed statefull test. There is dangerous bug.

**Comment by alesapin** [comment]
It's not 100% ready, I still can do some refactoring and optimizations. However I'd be glad to see some review notices.

**Comment by alesapin** [comment]
00653_verification_monotonic_data_load and 00700_decimal_round broken in master. 00806_alter_update in unbundled mode seems strange.

File: dbms/programs/compressor/Compressor.cpp [hunk 1]
```diff
@@ -85,18 +86,20 @@ int mainEntryClickHouseCompressor(int argc, char ** argv)
...
+            method_family = "NONE";
 
-        DB::CompressionSettings settings(method, options.count("level")
-            ? options["level"].as<int>()
-            : DB::CompressionSettings::getDefaultLevel(method));
+        std::optional<int> level;
+        if (options.count("level"))
+            level = options["level"].as<int>();
+
+        DB::CompressionCodecPtr codec = DB::CompressionCodecFactory::instance().get(method_family, level);
```
Comment (alexey-milovidov) [inline dbms/programs/compressor/Compressor.cpp]: You can also introduce `--codec` option that will allow to pass stacked codecs.

Comment (alesapin) [reply dbms/programs/compressor/Compressor.cpp]: https://github.com/yandex/ClickHouse/pull/3899/files#diff-238917e94163b6eba9b15c42a9aa2b55

**Pull request merged**